### ISSUE TITLE
Highlight other languages in macros and strings

### ISSUE THREAD

**Issue by adiba** [intellij-rust/intellij-rust#2231]
There are many possibilities to write inline scripts like dyno or ASM in Rust.
Example of JavaScript using stdweb:
```rust
js!( alert("here goes javascript") )
```
Highlighting and inspection of code in strings or macro bodies would really be helpful. This already exists for Java:
https://intellij-support.jetbrains.com/hc/en-us/community/posts/206899855-mixing-two-languages-in-one-file-

Related: https://github.com/intellij-rust/intellij-rust/issues/1940

**Comment by garyttierney** [comment]
Further on @vlad20012's comment, it'd be cool if those language injections could be user-defined using [IntelliLang](https://plugins.jetbrains.com/plugin/1105-intellilang) (e.g. I can create a selector that says the body of a macro invocation named _xyz_ is JavaScript, much like you can already do with XML injections.  Relevant code ).

I have a bit of a WIP patch that adds the necessary UI elements and PSI matching code for this that I can push up later to play around with.  The benefit of this approach is that you can also change/remove the regex injections without code changes.

**Comment by Undin** [comment]
@garyttierney `Inject language or reference` intention has worked for string literals since #2679
![2018-11-12 22 58 01](https://user-images.githubusercontent.com/2539310/48372078-84c5b100-e6ce-11e8-9929-0c59c45acc1c.gif)
Or do you mean something else?

**Comment by garyttierney** [comment]
@Undin yes, for Java, XML, Kotlin, and Groovy you can define code patterns where language injections will automatically happen.  It's not core IntelliJ functionality, but it's shipped as part of the IntelliLang plugin that is bundled by default.

![peek 2018-11-12 20-09](https://user-images.githubusercontent.com/1180094/48372554-dd894f80-e6b6-11e8-80ad-a9e7cf1cc0c9.gif)

**Comment by garyttierney** [comment]
Hm, you're right.  That's unfortunate.  I wonder if there's any reason it couldn't work in CLion.  I'll take a look at that.

**Comment by Undin** [comment]
@garyttierney FYI, the corresponding [issue](https://youtrack.jetbrains.com/issue/CPP-6790) in CLion issue tracker

**Comment by aldanor** [comment]
It would be also nice if $expr items in macros were treated (and indented) like proper Rust code.

**Comment by garyttierney** [comment]
I think this is now unblocked by the 2019.1 EAP of CLion (the YouTrack linked above has a fix version of CLion 2019.1 EAP (191.4212.38)).

**Comment by Undin** [comment]
@garyttierney Yep, now CLion has bundled IntelliLang plugin

**Comment by vlad20012** [comment]
@garyttierney Recently I looked at it the implementation of such patterns. It isn't to hard to implement, but we need to invent patterns for rust at least (Considering possible methods in impls for complex types). 
Also we're blocked on #2733 (As far as I know, someone from the Core is working on it now)

**Comment by vlad20012** [comment]
Ok, #2733 is now fixed in 2019.1

**Comment by garyttierney** [comment]
Great! I have one PR I need to finish then I will continue working on this if no one else intends on picking it up (I have a WIP that would only work in IDEA previously).

**Comment by NyxCode** [comment]
have you made any progress on this @garyttierney? 
I could really use this feature for inline SQL statements.

**Comment by vlad20012** [comment]
@NyxCode could you please list crates that has ability to write SQL statements as strings? I know that the most popular framework is `diesel` that is a kind of SQL builder.

**Comment by NyxCode** [comment]
@vlad20012 
`mysql` is the one i'm using, but i'm sure there are many more.
`postgres`, `tokio-postgres`, `oracle` - just to name a few.

**Comment by abonander** [comment]
We just launched SQLx which leans heavily on raw SQL strings: https://github.com/launchbadge/sqlx#examples

Fortunately, the injection rules should be pretty simple. If it would be accepted, I'd gladly work on a PR to add them.

**Comment by abonander** [comment]
Could we support a general comment-based syntax that the IntelliJ-Rust plugin could recognize and add language injections on the fly? E.g. with `sqlx`:

```rust
// @intellij-rust: lang=sql,match=$query
macro_rules! query (
    ($query:literal) => (...);
    ($query:literal, $($args:tt)*) => (...)
);

// @intellij-rust: lang=sql,match=sql
pub fn query<DB>(sql: &str) -> Query<DB>
where
    DB: Database,
{
    Query {
        database: PhantomData,
        arguments: Default::default(),
        query: sql,
    }
}

**Comment by ryanbrainard** [comment]
> Great! I have one PR I need to finish then I will continue working on this if no one else intends on picking it up (I have a WIP that would only work in IDEA previously).

@garyttierney Any progress on this? 

---

I was experimenting a bit with this and IntelliJ does support comments to hint at code injection. Like this:

```
// language=sql
let query = "SELECT id FROM users;"
```

Fortunately, this works on freestanding lines, but unfortunately, it doesn't work in `sqlx` macros because the argument must be a static string literal, which it no longer is when assigned to a variable. I tried putting the comment inline in the macro in a variety of ways, but couldn't get it to work. 

While I'd much rather have it _just work_, I'd be totally fine having to add comments like these to get the syntax highlighting. Just mentioning this in case the fix/implementation is easier to support this.

**Comment by vlad20012** [comment]
@ryanbrainard Hello! I can make auto injections for well-known SQL macros. Could you please list some known crates?

**Comment by ryanbrainard** [comment]
> @ryanbrainard Hello! I can make auto injections for well-known SQL macros/functions. Could you please list some known crates?

@vlad20012 That would be excellent! The biggest one I know of is [`sqlx`](https://crates.io/crates/sqlx) with it's `query!`, `query_as!`, `query()`, `query_as()`, etc. 

I also noticed with these macros, IntelliJ doesn't show the "Inject language or reference" option on the context (right-click) menu. Instead, it only shows the "Show ... macro expansion" options, so I wonder if the macro expansion is somehow hiding or overriding the default language injection option for strings.

**Comment by mehcode** [comment]
cc @vlad20012 @ryanbrainard 

Has there been any more progress on this? Is there an example of an auto-injection somewhere? I don't mind trying to send a PR for this.

**Comment by JarvisCraft** [comment]
> I was experimenting a bit with this and IntelliJ Rust does support comments to hint at code injection. Like this:
> 
> ```
> // language=sql
> let query = "SELECT id FROM users;"
> ```
> 
> Fortunately, this works on freestanding lines, but unfortunately, it doesn't work in `sqlx` macros because the argument must be a static string literal, which it no longer is when assigned to a variable. I tried putting the comment inline in the macro in a variety of ways, but couldn't get it to work.

Actually, this approach seems to work for literals themselves:

![image](https://user-images.githubusercontent.com/7693005/122604386-8e7dd500-d07e-11eb-9591-05c3825d4364.png)

**Comment by abonander** [comment]
Yep, we've been using this in our dogfooding usage of SQLx for some time now. It would be nice to not have to annotate every query string, though, which is why I would still like support for adding this annotation or something similar to the `query!()` macro definition so it just automatically works.

**Comment by xpe** [comment]
In Clion, I use `// language=sql` above the function containing an [`indoc!`](https://github.com/dtolnay/indoc) macro:

```rust
// language=sql
pub fn create_table(conn: &Connection) -> Result<usize> {
    let sql = indoc! {"
        create table account
        ( id         integer primary key
        , name       varchar unique not null
        , created_at integer primary key
        )"};
    conn.execute(sql, [])
}
```

This code uses the [rusqlite](https://github.com/rusqlite/rusqlite) SQLite library. Here is how it highlights in CLion:

<img width="535" alt="image" src="https://user-images.githubusercontent.com/4576329/130341308-a8ce59a4-5727-470a-8135-32d27878de70.png">

P.S. This also works with the magic-comment line right above the `let sql = ...` line.

**Comment by longfellowone** [comment]
Any updates on this? It seems to work with SQLx when hitting ALT+ENTER and manually injecting, just not automatically.

**Comment by JarvisCraft** [comment]
> Any updates on this? It seems to work with SQLx when hitting ALT+ENTER and manually injecting, just not automatically.

As said above, you can use comment-based language hints, e.g. `// language=sql` which is persistent.

**Comment by abonander** [comment]
I would like to know if there's someplace that SQLx can add `// language=sql` to its macro expansions to get this to work automatically, as I think that would be the easiest way to get this working.
